lib/instance.js

```diff
@@ -207,57 +207,119 @@ function instance(system) {
-					self.active[id].label = config.label
+		system.emit('instances_loaded')
+	})
 
-					if (typeof self.active[id].upgradeConfig == 'function') {
-						self.active[id].upgradeConfig()
+	system.on(
+		'instance_upgrade_imported',
+		function (instance_id, instance_created, instance_type, imported_config, actions, feedbacks) {
+			if (self.checkModuleLoaded(instance_type)) {
+				const mod = self.modules[instance_type]
+				if (mod.GetUpgradeScripts) {
+					const upgrade_scripts = mod.GetUpgradeScripts()
+
+					// config should only be upgraded if the instance was created from the import
+					const upgrade_config = instance_created ? instance_config : null
+					const instance_config = self.store.db[instance_id]
+
+					// If the instance was just created, we want to run everything. If it already exists, we want to match the version in the stored config
+					const targetIdx = instance_created
+						? upgrade_scripts.length
+						: Math.max(instance_config._configIdx + 1, upgrade_scripts.length)
+					const fromIdx = imported_config._configIdx ?? -1
+
+					if (idx + 1 < upgrade_scripts.length) {
+						debug('upgradeImported(' + instance_type + '): ' + (fromIdx + 1) + ' to ' + targetIdx)
 					}
 
-					if (typeof self.active[id]._init == 'function') {
-						debug('Running _init of ' + id)
+					let changed = false
+					for (var i = idx + 1; i < targetIdx; ++i) {
+						debug('UpgradeConfig: Upgrading to version ' + (i + 1))
+
 						try {
-							self.active[id]._init()
+							const result = upgrade_scripts[i](upgrade_config, actions, release_actions, feedbacks)
+							changed = changed || result
 						} catch (e) {
-							self.system.emit(
-								'log',
-								'instance(' + self.active[id].label + ')',
-								'warn',
-								'Error initalizing module: ' + e.message
-							)
+							debug('Upgradescript in ' + instance_type + ' failed', e)
+						}
+						instance_config._configIdx = i
+
+						for (const action of [...actions, release_actions]) {
+							action.instance = instance_id
+							action.label = `${instance_id}:${action.action}`
+						}
+						for (const feedback of feedbacks) {
+							feedback.instance_id = instance_id
 						}
 					}
-				} else {
-					debug('Configured instance ' + config.instance_type + ' could not be loaded, unknown module')
-					if (config.instance_type != 'bitfocus-companion') {
-						system.emit(
-							'log',
-							'instance(' + config.instance_type + ')',
-							'error',
-							'Configured instance ' + config.instance_type + ' could not be loaded, unknown module'
-						)
+
+					if (idx + 1 < targetIdx) {
+						// Save config, but do not automatically call this module's updateConfig again
+						// Either only _configIdx changed, or the module is not active yet
+						self.system.emit('instance_config_put', instance_id, instance_config, true)
 					}
+
+					// If anything was changed, update system and db
+					if (changed) {
+						system.emit('config_save')
```

Comment:
I know it was there before, but this shouldn't be needed ... its the sysconfig (ip, port, etc)